plugins/warpsync/component.go

```diff
@@ -61,109 +52,88 @@ type dependencies struct {
-	})
-
-	onWarpSyncStart = events.NewClosure(func(targetMsIndex iotago.MilestoneIndex, nextCheckpoint iotago.MilestoneIndex, advRange syncmanager.MilestoneIndexDelta) {
-		Plugin.LogInfof("Synchronizing to milestone %d", targetMsIndex)
-		deps.RequestQueue.Filter(func(r *gossip.Request) bool {
-			return r.MilestoneIndex <= nextCheckpoint
-		})
-
-		msRequested, _, _ := warpSyncMilestoneRequester.RequestMilestoneRange(advRange)
-		// if the amount of requested milestones doesn't correspond to the range,
-		// it means we already had the milestones in the database, which suggests
-		// that we should manually kick start the milestone solidifier.
-		if msRequested != advRange {
-			Plugin.LogInfo("Manually starting solidifier, as some milestones are already in the database")
-			deps.Tangle.TriggerSolidifier()
-		}
-	})
-
-	onWarpSyncDone = events.NewClosure(func(deltaSynced int, referencedBlocksTotal int, took time.Duration) {
-		// we need to cleanup all memoized things in the requester, so we have a clean state at next run and free the memory.
-		// we can only reset the "traversed" blocks here, because otherwise it may happen that the requester always
-		// walks the whole cone if there are already paths between newer milestones in the database.
-		warpSyncMilestoneRequester.Cleanup()
-
-		Plugin.LogInfof("Synchronized %d milestones in %v (%0.2f BPS)", deltaSynced, took.Truncate(time.Millisecond), float64(referencedBlocksTotal)/took.Seconds())
-		deps.RequestQueue.Filter(nil)
-	})
-}
-
-func attachEvents() {
-	deps.GossipService.Events.ProtocolStarted.Hook(onGossipServiceProtocolStarted)
-	deps.GossipService.Events.ProtocolTerminated.Hook(onGossipServiceProtocolTerminated)
-	deps.Tangle.Events.ReferencedBlocksCountUpdated.Hook(onReferencedBlocksCountUpdated)
-	deps.Tangle.Events.MilestoneSolidificationFailed.Hook(onMilestoneSolidificationFailed)
-	warpSync.Events.CheckpointUpdated.Hook(onWarpSyncCheckpointUpdated)
-	warpSync.Events.TargetUpdated.Hook(onWarpSyncTargetUpdated)
-	warpSync.Events.Start.Hook(onWarpSyncStart)
-	warpSync.Events.Done.Hook(onWarpSyncDone)
-}
-
-func detachEvents() {
-	deps.GossipService.Events.ProtocolStarted.Detach(onGossipServiceProtocolStarted)
-	deps.GossipService.Events.ProtocolTerminated.Detach(onGossipServiceProtocolTerminated)
-	deps.Tangle.Events.ReferencedBlocksCountUpdated.Detach(onReferencedBlocksCountUpdated)
-	deps.Tangle.Events.MilestoneSolidificationFailed.Detach(onMilestoneSolidificationFailed)
-	warpSync.Events.CheckpointUpdated.Detach(onWarpSyncCheckpointUpdated)
-	warpSync.Events.TargetUpdated.Detach(onWarpSyncTargetUpdated)
-	warpSync.Events.Start.Detach(onWarpSyncStart)
-	warpSync.Events.Done.Detach(onWarpSyncDone)
+func hookEvents() (detach func()) {
+	var heartbeakUnhook func()
+	return lo.Batch(
+		deps.GossipService.Events.ProtocolStarted.Hook(func(p *gossip.Protocol) {
+			heartbeakUnhook = p.Events.HeartbeatUpdated.Hook(func(hb *gossip.Heartbeat) {
```

Comment:
this seems wrong to me on the first look.
this heartbeat unhook is only valid for this `gossip.Protocol`, but is reused for every other `ProtocolTerminated` event.